tools/clang/test/HLSL/atomic-float-errors.hlsl

```diff
@@ -0,0 +1,55 @@
+// RUN: %clang_cc1 -fsyntax-only -ffreestanding -verify  %s
+
+// Verify that the first arg determines the overload and the others can be what they will
+
+groupshared int      resGI[256];
+groupshared uint64_t resGI64[256];
+RWBuffer<int>      resBI;
+RWBuffer<uint64_t> resBI64;
+
+RWByteAddressBuffer Rres;
+
+[numthreads(1,1,1)]
+void main( uint3 gtid : SV_GroupThreadID)
+{
+  uint a = gtid.x;
+  uint b = gtid.y;
+  uint c = gtid.z;
+  resGI[a] = a;
+  resGI64[a] = a;
+  resBI[a] = a;
+  resBI64[a] = a;
+
+  float fv = b - c;
+  float fv2 = b + c;
+  float ofv = 0;
+  int iv = b / c;
+  int iv2 = b * c;
+  int oiv = 0;
+  uint64_t bb = b;
+  uint64_t cc = c;
+  uint64_t lv = bb * cc;
+  uint64_t lv2 = bb / cc;
+  uint64_t olv = 0;
+
+  InterlockedCompareStoreFloatBitwise( resBI[a], iv, iv2 ); // expected-error{{no matching function for call to 'InterlockedCompareStoreFloatBitwise'}} expected-note{{candidate function not viable: no known conversion from 'int' to 'float &' for 1st argument}}
```

Comment:
Sorry for the basic question, but does this test atomics with and without a return value? Or is the 'ref' aspect only valid for atomics with return?

Reply:
That’s a good question. This test case is non-exhaustive, but it just verifies that overload resolution is properly generating a mismatch on the first parameter. There are other test cases (`atomic_overload.hlsl`, `atomic_restypes.hlsl`, and `atomicop_float.hlsl`) which verify the same error-generation for other atomic functions.

This is showing up in the change as a new file, but it is just a rewritten test from being FileCheck-based to being Clang verifier based. Making it verifier based means the source location of the diagnostics also get verified. In general we shouldn’t be using FileCheck for verifying diagnostics because of the lack of verifying source locations.

For a little more context: This change is more or less NFC to the user. The code I deleted from SemaHLSL.cpp matched the intrinsic opcode then treated the first parameter of those intrinsics as a reference. This change effectively moves that logic out of special casing for overload resolution and function generation, and instead captures it where we specify the function signature so we just generate the signature correctly up front and everything else mostly just works as expected.

Reply:
Yea. I didn't mark the change NFC because it does produce some slightly different diagnostics (hence the updated tests), and it produces some verifiable AST differences.

Thank you for the through review 😀